File: python/08_048.py
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def interleave_lists(lists):
    if not lists:
        return None
    
    # Remove any empty lists
    lists = [lst for lst in lists if lst]
    if not lists:
        return None
    
    dummy = ListNode(0)
    tail = dummy
    
    while lists:
        # Process one node from each list in round-robin fashion
        for i in range(len(lists)):
            # Append current node to result
            tail.next = lists[i]
            tail = tail.next
            
            # Move to next node in current list
            lists[i] = lists[i].next
        
        # Remove any lists that became empty
        lists = [lst for lst in lists if lst]
    
    return dummy.next

File: data structures and algorithms/cs/08_048.cs
using System.Collections.Generic;

public class ListNode {
    public int Val { get; set; }
    public ListNode Next { get; set; }

    public ListNode(int val) {
        Val = val;
    }
}

public class Solution {
    public ListNode InterleaveLists(ListNode[] lists) {
        if (lists == null || lists.Length == 0) {
            return null;
        }

        var activeList = new List<ListNode>();
        foreach (var list in lists) {
            if (list != null) {
                activeList.Add(list);
            }
        }

        if (activeList.Count == 0) {
            return null;
        }

        ListNode dummy = new ListNode(0);
        ListNode tail = dummy;

        while (activeList.Count > 0) {
            int size = activeList.Count;
            for (int i = 0; i < size; i++) {
                ListNode current = activeList[0];
                activeList.RemoveAt(0);

                tail.Next = current;
                tail = tail.Next;

                if (current.Next != null) {
                    activeList.Add(current.Next);
                }
            }
        }

        return dummy.Next;
    }
}